File: NatureOfCode in Swift(Retired)/NatureOfCode/Walker.swift
//
//  Walker.swift
//  NatureOfCode
//
//  Created by Adrian Simionescu on 24/09/14.
//  Copyright (c) 2014 Adrian Simionescu. All rights reserved.
//

import Foundation


/**
A protocol to conform to the class that implements the Walker class.
*/
protocol PWalker : PObjectBaseBasic
{
    
    /**
    The implementor of the Walker class must have an Walker object
    */
    var WalkerObject : Walker { get set }
    
    /**
    Implement this method to render the walker at a certain position
    */
    func RenderWalkerByPosition( position: CGPoint);
    
    
    /**
    Use this method to set a target for the Walker to aim at, to move towards to
    */
    func SetWalkerTarget(target : CGPoint);
    
    /**
    Use this method to implement logic related to make the Walker to give you the next position where the implementor should be at
    */
    mutating func StepWalker();
    
    /**
    Use this method to implement logic related to make the Walker to give you the next position where the implementor should be at and aimed at the target, moving towardst he target.
    */
    mutating func StepWalkerByTarget(target : CGPoint);
}

// Walker base class that gives functionality to some object. Use it to move it randomly somewhere
public class Walker : RandomBaseComplete
{
    /// Current X position
    var PositionX : CGFloat = 0;
    
    /// Current Y position
    var PositionY : CGFloat = 0;
    
    var Target : CGPoint = CGPointZero;
    
    var PerlinNoiseTime_PositionX : CGFloat = 0;
    var PerlinNoiseTime_PositionY : CGFloat = 10000;

    
    /// Probality factor walking towards a target
    var ProbalitityFactor : Float = 0.9;
    
    
    init(width : CGFloat, height : CGFloat)
    {
        super.init();
        self.PositionX = width / 2;
        self.PositionY = height / 2;
        self.Width = width;
        self.Height = height;
        self.RandomNumberModeValue = RandomNumberMode.Uniform;
    }
    
    init(positionX : CGFloat, positionY : CGFloat, width : CGFloat, height : CGFloat)
    {
        super.init();
        self.PositionX = positionX;
        self.PositionY = positionY;
        self.Width = width;
        self.Height = height;
        self.RandomNumberModeValue = RandomNumberMode.Uniform;
    }
    
    /// Set a probability factor when to start to walk towards the target
    func SetProbalitiyFactor(factor : Float)
    {
        self.ProbalitityFactor = factor;
    }
    
    override func UniformCalculations()
    {
        super.UniformCalculations();
        if( self.Target == CGPointZero)
        {
            var choice = RandomNumberGenerator.GetRandomUInt32(4);
        
            switch(choice)
                {
            case 0:
                self.PositionX += RandomNumberGenerator.GetRandomCGFloat(2);
            case 1:
                self.PositionX -= RandomNumberGenerator.GetRandomCGFloat(2);
            case 2:
                self.PositionY += RandomNumberGenerator.GetRandomCGFloat(2);
            case 3:
                self.PositionY -= RandomNumberGenerator.GetRandomCGFloat(2);
            default:
                self.PositionX++;
            }
        } else
        {
            MoveWalkerTowardsTarget();
        }
    }
    
    func MoveWalkerTowardsTarget()
    {
        var choice : Float = RandomNumberGenerator.GetRandomPositiveFloat(1);
        
        if(choice < self.ProbalitityFactor)
        {
            var xdir = (self.Target.x - self.PositionX);
            var ydir = (self.Target.y - self.PositionY);
            
            if(xdir != 0)
            {
                xdir /= abs(xdir);
            }
            
            if(ydir != 0)
            {
                ydir /= abs(ydir);
            }
            
            self.PositionX += xdir;
            self.PositionY += ydir;
        } else
        {
            self.PositionX += RandomNumberGenerator.GetRandomCGFloat(2);
            self.PositionY += RandomNumberGenerator.GetRandomCGFloat(2);
        }
    }
    
    override func GaussianCalculations()
    {
        super.GaussianCalculations();
    }
    
    override func PerlinNoiseCalculations()
    {
        super.PerlinNoiseCalculations();
        
        
        var tempX = CGFloat(self.PerlinNoiseCalculator.perlin1DValueForPoint(Float(self.PerlinNoiseTime_PositionX)));
        var tempY = CGFloat(self.PerlinNoiseCalculator.perlin1DValueForPoint(Float(self.PerlinNoiseTime_PositionY)));
        
        if(tempX < 0)
        {
            tempX *= -1;
        }
        
        if(tempY < 0)
        {
            tempY *= -1;
        }
        self.PositionX = PerlinNoiseTool.MapCGFloat(tempX, perlinNoiseMinValue: 0, perlinNoiseMaxValue: 1, targetMinValue: 0, targetMaxValue: self.Width);
        
        self.PositionY = PerlinNoiseTool.MapCGFloat(tempY, perlinNoiseMinValue: 0, perlinNoiseMaxValue: 1, targetMinValue: 0, targetMaxValue: self.Height);
        
        self.PerlinNoiseTime_PositionX += 0.1;
        self.PerlinNoiseTime_PositionY += 0.1;
        
        if( self.Target != CGPointZero)
        {
            self.MoveWalkerTowardsTarget();
        }
    }
    
    /// Makes the walker object to take a step in a direction
    /*public func step() -> CGPoint
    {
        self.target = CGPointZero;
        self.UniformCalculations();
        
        
        self.constrain();
        
        return(CGPointMake(self.positionX, self.positionY));
    }*/
    
    public func SetTarget(target : CGPoint)
    {
        self.Target = target;
    }
    
    /// Makes the walker object to take a step towards the target
    /*public func step(target : CGPoint) -> CGPoint
    {
        self.Target = target;
        self.UniformCalculations();
        
        self.constrain();
        
        return(CGPointMake(self.positionX, self.positionY));
    }*/
    
    /// Return current position of the walker
    public func GetCurentPosition() -> CGPoint
    {
        return(CGPointMake(self.PositionX, self.PositionY));
    }
    
    /// Constrain the walker position to the defined area
    override func Constrain()
    {
        super.Constrain();
        if(self.PositionX < 0)
        {
            self.PositionX = 0;
        }
        if(self.PositionX > self.Width)
        {
            self.PositionX = self.Width;
        }
        if(self.PositionY < 0)
        {
            self.PositionY = 0;
        }
        
        if(self.PositionY > self.Height)
        {
            self.PositionY = self.Height;
        }
        
    }
}

File: NatureOfCodeCocos2DX/Classes/Aurora/Random/Implementations/Walker.cpp
#include "Walker.h"

namespace Aurora {
	namespace Random {

		void Walker::Constrain()
		{
			RandomBaseComplete::Constrain();
			if(this->position.X < 0)
			{
				this->position.X = 0;
			}
			if(this->position.X > this->GetConstrainsRange().Width)
			{
				this->position.X = this->GetConstrainsRange().Width;
			}
			if(this->position.Y < 0)
			{
				this->position.Y = 0;
			}

			if(this->position.Y > this->GetConstrainsRange().Height)
			{
				this->position.Y = this->GetConstrainsRange().Height;
			}
		}

		void Walker::UniformCalculations()
		{
			RandomBaseComplete::UniformCalculations();
			VECTOR2D emptyVector;

			if(this->target == emptyVector)
			{
				auto choice = static_cast<int>(RandomNumberGenerator::GetRandomPositiveFloat(0,4));
				auto randomMovementMagnitude = RandomNumberGenerator::GetRandomPositiveFloat(4);
				switch(choice)
				{
				case 0:
				{
						  this->position.X += randomMovementMagnitude;
				} break;
				case 1:
				{
						  this->position.Y += randomMovementMagnitude;
				} break;
				case 2:
				{
						  this->position.X -= randomMovementMagnitude;
				} break;
				case 3:
				{
						  this->position.Y -= randomMovementMagnitude;
				} break;
				default:
				{
						   this->position.X++;
				} break;
				}
			} else
			{
				MoveWalkerTowardsTarget();
			}
		}

		void Walker::GaussianCalculations()
		{
			RandomBaseComplete::GaussianCalculations();
		}

		void Walker::PerlinNoiseCalculations()
		{
			RandomBaseComplete::PerlinNoiseCalculations();
			auto tempX = this->perlinNoiseCalculator->GetValue(this->perlinNoiseTime_PositionX.X, this->perlinNoiseTime_PositionX.Y, this->perlinNoiseTime_PositionX.Z);
			auto tempY = this->perlinNoiseCalculator->GetValue(this->perlinNoiseTime_PositionY.X, this->perlinNoiseTime_PositionY.Y, this->perlinNoiseTime_PositionY.Z);

			if(tempX < 0)
			{
				tempX *= -1;
			}

			if(tempY < 0)
			{
				tempY *= -1;
			}
			
			this->position.X = PerlinNoiseTool::Map(tempX, 0, 1, 0, this->GetConstrainsRange().Width);
			this->position.Y = PerlinNoiseTool::Map(tempY, 0, 1, 0, this->GetConstrainsRange().Height);

			this->perlinNoiseTime_PositionX += 0.1f;
			this->perlinNoiseTime_PositionY += 0.1f;

			VECTOR2D emptyVector;

			if(this->target != emptyVector)
			{
				this->MoveWalkerTowardsTarget();
			}
		}

		void Walker::NormalCalculations()
		{
			RandomBaseComplete::NormalCalculations();
		}

		Walker::Walker(const mRECT &areaSize) : RandomBaseComplete()
		{
			this->init(areaSize, VECTOR2D::GetZeroVector());
		}

		Walker::Walker(const mRECT &areaSize, const VECTOR2D &walkerStartPosition) : RandomBaseComplete(), position(walkerStartPosition)
		{
			this->init(areaSize, walkerStartPosition);
		}

		Walker::Walker() : RandomBaseComplete()
		{
			this->init();
		}

		Walker::Walker(const Walker &value) : RandomBaseComplete(value)
		{
			this->init(value);
		}

		Walker::Walker(Walker &&value) : RandomBaseComplete(std::move(value))
		{
			this->position = std::move(value.position);
			this->probalitityFactor = value.probalitityFactor;
			this->perlinNoiseTime_PositionX = std::move(value.perlinNoiseTime_PositionX);
			this->perlinNoiseTime_PositionY = std::move(value.perlinNoiseTime_PositionY);
			this->target = std::move(value.target);
		}


		Walker::~Walker()
		{
			RandomBaseComplete::~RandomBaseComplete();
			IPhysicsImplementor::~IPhysicsImplementor();
		}

		void Walker::MoveWalkerTowardsTarget()
		{
			Float choice = RandomNumberGenerator::GetRandomPositiveFloat(0,1);

			if(choice < this->probalitityFactor)
			{
				auto xdir = (this->target.X - this->position.X);
				auto ydir = (this->target.Y - this->position.Y);

				if(xdir != 0)
				{
					xdir /= abs(xdir);
				}

				if(ydir != 0)
				{
					ydir /= abs(ydir);
				}

				this->position.X += xdir;
				this->position.Y += ydir;
			} else
			{
				this->position.X += RandomNumberGenerator::GetRandomFloat(2);
				this->position.Y += RandomNumberGenerator::GetRandomFloat(2);
			}
			this->target = VECTOR2D::GetZeroVector();
		}

		void Walker::SetTarget(const VECTOR2D &target)
		{
			this->target = target;
		}

		const VECTOR2D  Walker::GetCurentPosition() const
		{
			return(this->position);
		}

		void Walker::SetProbalitiyFactor(Float probalitityFactor)
		{
			this->probalitityFactor = probalitityFactor;
		}

		void Walker::SetPosition(const VECTOR2D &position)
		{
			this->position = position;
		}

		Walker& Walker::operator=(const Walker& value)
		{
			if(this == &value) { return(*this); }
			RandomBaseComplete::operator=(value);

			this->init(value);
			
			return(*this);
		}

		Walker & Walker::operator=(Walker && value)
		{
			if (this == &value)
				return *this;

			
			this->position = std::move(value.position);
			this->probalitityFactor = value.probalitityFactor;
			this->perlinNoiseTime_PositionX = std::move(value.perlinNoiseTime_PositionX);
			this->perlinNoiseTime_PositionY = std::move(value.perlinNoiseTime_PositionY);
			this->target = std::move(value.target);
			
			RandomBaseComplete::operator=(std::move(value));
			return(*this);
		}

		void Walker::init()
		{
			this->SetConstrainsRange(mRECT(0,0));
			this->target.ZeroVector();
			this->perlinNoiseTime_PositionX = VECTOR3D(0,0,0);
			this->perlinNoiseTime_PositionY = VECTOR3D(10000,10000,10000);
			this->probalitityFactor = 0.9f;
			this->position.ZeroVector();
			this->SetRandomNumberMode(RandomNumberMode::Uniform);
		}

		void Walker::init(const Walker &value)
		{
			this->SetConstrainsRange(value.GetConstrainsRange());
			//this->PerlinNoiseCalculator = value.PerlinNoiseCalculator;
			this->position = value.position;
			this->probalitityFactor = value.probalitityFactor;
			this->SetRandomNumberMode(value.GetRandomNumberMode());
			this->perlinNoiseTime_PositionX = value.perlinNoiseTime_PositionX;
			this->perlinNoiseTime_PositionY = value.perlinNoiseTime_PositionY;
			this->target = value.target;
		}

		void Walker::init(const mRECT &areaSize, const VECTOR2D &walkerStartPosition)
		{
			this->target.ZeroVector();
			this->perlinNoiseTime_PositionX = VECTOR3D(0,0,0);
			this->perlinNoiseTime_PositionY = VECTOR3D(10000,10000,10000);
			this->probalitityFactor = 0.9f;

			if(!areaSize.IsZero())
				this->SetConstrainsRange(areaSize);

			if(!walkerStartPosition.IsZero())
				this->position = walkerStartPosition;

			this->SetRandomNumberMode(RandomNumberMode::Uniform);
		}


		void IWalkerImplementor::Render()
		{

		}

		void IWalkerImplementor::SetWalkerRandomNumberMode(RandomNumberMode randomNumberMode)
		{

		}

		void IWalkerImplementor::RenderWalkerByPosition(const VECTOR2D &position)
		{

		}

		void IWalkerImplementor::SetWalkerTarget(const VECTOR2D &target)
		{

		}

		void IWalkerImplementor::StepWalker()
		{

		}

		std::shared_ptr<Physics::Force> IWalkerImplementor::AccessObjectPhysics() const
		{
			return nullptr;
		}


	}; // END OF NAMESPACE Random
}; // END OF NAMESPACE Aurora